### PR TITLE
[IR] Add vector type support for program translator

### PR DESCRIPTION
**Pull request by kangguangli**
### PR types

New features
### PR changes

Others
### Description

In #53707 , we developed a simple program translator, which translates op's input/output, which type is `Map<Str, Vec<Str>>`, into `Vec<Value>`.  The logic is close to following code:
```
new_inputs = concat(old_inputs.values())
```
However, it does not match the function signature in `phi` operator library and does not satisfy the semantics of old representation. The input/output of `OpDesc` have two types:
1. Tensor: corresponding a vector<Str> with size == 1
2. Vector\<Tensor\>: corresponding a vector<Str> with size > 1

So we need to distinguish these two types. Now we add two builtin ops `combine` and `slice` to help finish these. After this PR merges, the translation result will be like:
```
(%1273)  = "builtin.combine" (%1255,%1272)  { ATTRIBUTE }  : tensor<-1x64x56x56xf32>,tensor<-1x64x56x56xf32> -> vec<tensor<-1x64x56x56xf32>,tensor<-1x64x56x56xf32>>
(%1274)  = "pd.sum" (%1273)  { ATTRIBUTE }  : vec<tensor<-1x64x56x56xf32>,tensor<-1x64x56x56xf32>> -> tensor<-1x64x56x56xf32>
```

#### Others
Pcard-67164

### CODE REVIEW

File: paddle/ir/builtin_dialect.cc [hunk 1]
```diff
@@ -44,7 +44,10 @@ void BuiltinDialect::initialize() {
                      ir::Int64_tAttribute,
                      ir::ArrayAttribute>();
 
-  RegisterOps<ir::GetParameterOp, ir::SetParameterOp>();
+  RegisterOps<ir::GetParameterOp,
```
Comment (winter-wang) [inline paddle/ir/builtin_dialect.cc:47]: 同在ir命名空间下，就没有使用ir::来标志了吧。

File: paddle/ir/builtin_op.cc [hunk 2]
```diff
@@ -58,4 +58,14 @@ void SetParameterOp::verify(const std::vector<ir::OpResult> &inputs,
   }
 }
 
+const char **CombineOp::attributes_name = nullptr;
+void CombineOp::verify(const std::vector<ir::OpResult> &inputs,
+                       const std::vector<ir::Type> &outputs,
+                       const ir::AttributeMap &attributes) {}
```
Comment (zhangbo9674) [inline paddle/ir/builtin_op.cc]: Combine 和 Slice 的 verify：是不是要针对输入输出，对 VectorType 中持有的Type个数及类型进行校验？

Comment (kangguangli) [reply paddle/ir/builtin_op.cc]: Done.

**Review by winter-wang — APPROVED** [review]
LGTM

File: paddle/ir/builtin_dialect.cc (hunk 1 above)
Comment (kangguangli) [reply paddle/ir/builtin_dialect.cc:47]: 好的，下个PR改一下。

File: paddle/fluid/translator/op_translator.cc [hunk 3]
```diff
@@ -84,23 +86,101 @@ inline ir::OpInfo LoopkUpOpInfo(ir::IrContext* ctx, const OpDesc& op_desc) {
...
+    ir::IrContext* ctx,
+    TranslationContext* param_map,
+    ir::Program* program,
+    const std::vector<std::string>& args) {
+  std::string combine_op_name(ir::CombineOp::name());
+  ir::OpInfo op_info = ctx->GetRegisteredOpInfo(combine_op_name);
+
+  std::vector<ir::OpResult> src_values;
+  std::vector<ir::Type> types_in_vec;
+  for (auto arg_name : args) {
```
Comment (Shixiaowei02) [inline paddle/fluid/translator/op_translator.cc:123]: `for (const auto& arg_name : args) { ... }`

Comment (kangguangli) [reply paddle/fluid/translator/op_translator.cc:123]: 好的，下个PR改一下。

File: paddle/fluid/translator/op_translator.cc [hunk 4]
```diff
@@ -84,23 +86,101 @@ inline ir::OpInfo LoopkUpOpInfo(ir::IrContext* ctx, const OpDesc& op_desc) {
...
+  return operation;
+}
+
 inline std::vector<ir::OpResult> GenerateOperationInput(
-    TranslationContext* param_map, const OpDesc& op_desc) {
+    ir::IrContext* ctx,
+    TranslationContext* param_map,
+    ir::Program* program,
+    const OpDesc& op_desc) {
   std::vector<ir::OpResult> op_inputs = {};
```
Comment (Shixiaowei02) [inline paddle/fluid/translator/op_translator.cc:140]: `= {}` 对 `std::vector` 是冗余的

File: paddle/fluid/translator/program_translator.h [hunk 5]
```diff
@@ -27,7 +27,25 @@
...
+                       int idx_in_vector = -1)
+      : value(value),
+        generated_by_vector(generated_by_vector),
+        idx_in_vector(idx_in_vector) {}
+  VariableDefiningInfo() {}
+
+  ir::OpResult value;
+
+  bool generated_by_vector =
+      false;  // true if target variabe is generated by Vector<Tensor>
```
Comment (Shixiaowei02) [inline paddle/fluid/translator/program_translator.h:42]: `target variable`

**Review by zhangbo9674 — APPROVED** [review]
LGTM

File: paddle/fluid/translator/program_translator.h (hunk 5 above)
Comment (kangguangli) [reply paddle/fluid/translator/program_translator.h:42]: 好的，下个PR改一下。

File: paddle/fluid/translator/op_translator.cc (hunk 4 above)
Comment (kangguangli) [reply paddle/fluid/translator/op_translator.cc:140]: 好的，下个PR改一下。

**Pull request merged**